flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/common/CommonExecSinkITCase.java

```diff
@@ -187,6 +198,148 @@ public void testStreamRecordTimestampInserterNotApplied() {
+
+            result = tableEnv.executeSql("SELECT * FROM T1");
+            result.await();
+
+            final List<String> expectedTrimmed =
+                    Arrays.asList(
+                            "+I[1, Apache F, SQL R, 11, 111, SQL]",
+                            "+I[2, Apache, SQL, 22, 222, Flink]",
+                            "+I[3, Apache, Flink, 33, 333, Apache]",
+                            "+I[4, Flink Pr, SQL o, 44, 444, Apache]");
+            final List<String> resultsAsStringStrimmed = new ArrayList<>();
+            result.collect().forEachRemaining(r -> resultsAsStringStrimmed.add(r.toString()));
+            assertEquals(expectedTrimmed, resultsAsStringStrimmed);
+
+        } finally {
+            tableEnv.getConfig()
+                    .getConfiguration()
+                    .setString(
+                            TABLE_EXEC_SINK_CHAR_PRECISION_ENFORCER.key(),
+                            ExecutionConfigOptions.CharPrecisionEnforcer.IGNORE.name());
+        }
+    }
+
+    @Test
+    public void testNullEnforcer() throws ExecutionException, InterruptedException {
+        final StreamTableEnvironment tableEnv = StreamTableEnvironment.create(env);
+        final List<Row> rows =
+                Arrays.asList(
+                        Row.of(1, "Apache", 11),
+                        Row.of(2, null, 22),
+                        Row.of(null, "Flink", 33),
+                        Row.of(null, null, 44));
+
+        final SharedReference<List<RowData>> results = sharedObjects.add(new ArrayList<>());
+        tableEnv.createTable(
+                "T1",
+                TableFactoryHarness.newBuilder()
+                        .schema(schemaForNotNullEnforcer())
+                        .source(new TestSource(rows))
+                        .sink(
+                                new TableFactoryHarness.SinkBase() {
+                                    @Override
+                                    public DynamicTableSink.SinkRuntimeProvider
+                                            getSinkRuntimeProvider(
+                                                    DynamicTableSink.Context context) {
+                                        return SinkProvider.of(
+                                                TestSink.newBuilder()
+                                                        .setWriter(new TestNotNullWriter(results))
+                                                        .setCommittableSerializer(
+                                                                TestSink.StringCommittableSerializer
+                                                                        .INSTANCE)
+                                                        .build());
+                                    }
+                                })
+                        .build());
+
+        // Default config - ignore (no trim)
+        ExecutionException ee =
```

Comment:
(This would / will be nicer with AssertJ… :upside_down_face: )

Reply:
indeed!

Reply:
indeed!